src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToComponentBehindCodeActionProvider.cs

```diff
@@ -0,0 +1,119 @@
+        if (syntaxTree?.Root is null)
+        {
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        var owner = syntaxTree.Root.FindInnermostNode(context.Location.AbsoluteIndex, true);
+        if (owner is null)
+        {
+            _logger.LogWarning($"Owner should never be null.");
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        var componentNode = owner?.FirstAncestorOrSelf<MarkupElementSyntax>();
+
+        // Make sure we've found tag
+        if (componentNode is null || componentNode.Kind != SyntaxKind.MarkupElement)
+        {
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        // Do not provide code action if the cursor is not inside a tag
+        if (context.Location.AbsoluteIndex > componentNode.StartTag.Span.End &&
+            context.Location.AbsoluteIndex < componentNode.EndTag.SpanStart)
+        {
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        if (!TryGetNamespace(context.CodeDocument, out var @namespace))
+        {
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        var actionParams = new ExtractToComponentBehindCodeActionParams()
+        {
+            Uri = context.Request.TextDocument.Uri,
+            ExtractStart = componentNode.Span.Start,
+            ExtractEnd = componentNode.Span.End,
+            Namespace = @namespace
+        };
+
+        var resolutionParams = new RazorCodeActionResolutionParams()
+        {
+            Action = LanguageServerConstants.CodeActions.ExtractToComponentBehindAction,
+            Language = LanguageServerConstants.CodeActions.Languages.Razor,
+            Data = actionParams,
+        };
+
+        var codeAction = RazorCodeActionFactory.CreateExtractToComponentBehind(resolutionParams);
+        var codeActions = new List<RazorVSInternalCodeAction> { codeAction };
+
+        return Task.FromResult<IReadOnlyList<RazorVSInternalCodeAction>?>(codeActions);
+    }
+
+    private static bool TryGetNamespace(RazorCodeDocument codeDocument, [NotNullWhen(returnValue: true)] out string? @namespace)
+        // If the compiler can't provide a computed namespace it will fallback to "__GeneratedComponent" or
+        // similar for the NamespaceNode. This would end up with extracting to a wrong namespace
+        // and causing compiler errors. Avoid offering this refactoring if we can't accurately get a
+        // good namespace to extract to
+        => codeDocument.TryComputeNamespace(fallbackToRootNamespace: true, out @namespace);
+
+    private static bool HasUnsupportedChildren(Language.Syntax.SyntaxNode node)
+    {
+        return node.DescendantNodes().Any(n =>
+            n is MarkupBlockSyntax ||
+            n is CSharpTransitionSyntax ||
+            n is RazorCommentBlockSyntax);
```

Comment:
nit: (nit stands for nitpick)

```suggestion
            n is MarkupBlockSyntax or
                  CSharpTransitionSyntax or
                  RazorCommentBlockSyntax);
```

This is a newer csharp feature than when some of the code was written, but I find it easier to read :)